src/transport/grpc/mod.rs

```diff
@@ -0,0 +1,1003 @@
+            .map(|_| ())?;
+
+        *stream = Grpc::attach(&mut client, data).await?;
+
+        Ok(())
+    }
+}
+
+impl std::fmt::Debug for Grpc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Grpc")
+            .field("uuid", &self.uuid)
+            .finish_non_exhaustive()
+    }
+}
+
+#[async_trait]
+impl Publish for Grpc {
+    async fn send_individual(&self, data: ValidatedIndividual<'_>) -> Result<(), crate::Error> {
+        self.client
+            .lock()
+            .await
+            .send(tonic::Request::new(
+                data.try_into().map_err(GrpcTransportError::from)?,
+            ))
+            .await
+            .map(|_| ())
+            .map_err(|e| GrpcTransportError::from(e).into())
+    }
+
+    async fn send_object(&self, data: ValidatedObject<'_>) -> Result<(), crate::Error> {
+        self.client
+            .lock()
+            .await
+            .send(tonic::Request::new(
+                data.try_into().map_err(GrpcTransportError::from)?,
+            ))
+            .await
+            .map(|_| ())
+            .map_err(|e| GrpcTransportError::from(e).into())
+    }
+}
+
+impl Deref for Grpc {
+    type Target = SharedGrpc;
+
+    fn deref(&self) -> &Self::Target {
+        &self.shared
+    }
+}
+
+#[async_trait]
+impl Receive for Grpc {
+    type Payload = GrpcReceivePayload;
+
+    async fn next_event<S>(
+        &self,
+        device: &SharedDevice<S>,
+    ) -> Result<ReceivedEvent<Self::Payload>, crate::Error>
+    where
+        S: PropertyStore,
+    {
+        loop {
+            match self.next_message().await {
+                Ok(Some(message)) => {
+                    let event: ReceivedEvent<Self::Payload> =
+                        message.try_into().map_err(GrpcTransportError::from)?;
+
+                    return Ok(event);
+                }
+                Ok(None) => {
+                    // try reattaching
+
+                    let data =
+                        NodeData::try_from_unlocked(self.uuid, &*device.interfaces.read().await)?;
+
+                    let mut stream = self.stream.lock().await;
+                    let mut client = self.client.lock().await;
+
+                    *stream = Grpc::attach(&mut client, data).await?;
+                }
+                Err(s)
+                    if s.code() == tonic::Code::Unavailable || s.code() == tonic::Code::Unknown =>
+                {
+                    // try reattaching
+
+                    let data =
+                        NodeData::try_from_unlocked(self.uuid, &*device.interfaces.read().await)?;
+
+                    let mut stream = self.stream.lock().await;
+                    let mut client = self.client.lock().await;
+
+                    *stream = Grpc::attach(&mut client, data).await?;
+                }
+                Err(s) => {
+                    return Err(GrpcTransportError::from(s).into());
+                }
+            }
```

Comment:
What do you think to merge `Ok(none)` and `Err(s) Unavailable & Unknown` statement? Something like that
```suggestion
                Err(s)
                    if s.code() != tonic::Code::Unavailable && s.code() != tonic::Code::Unknown =>
                {
                    return Err(GrpcTransportError::from(s).into());
                }
                _ => {
                   // try reattaching

                    let data =
                        NodeData::try_from_unlocked(self.uuid, &*device.interfaces.read().await)?;

                    let mut stream = self.stream.lock().await;
                    let mut client = self.client.lock().await;

                    *stream = Grpc::attach(&mut client, data).await?;
                }
            }
```

Reply:
Yes, that's better.